File: q1-q99/q10-q19/q15/q15.cpp
class Solution {
public:
     vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>>v;
        
        if(nums.size() < 3)
            return v;
        bool flag=true;
        for(int i=0; i<nums.size(); i++){
            if(nums[i] != 0){
                flag=false;
                break;
            }
        }
        if(flag){
            v.push_back({0, 0, 0});
            return v;
        }
        
        set<vector<int>>s;
        sort(nums.begin(), nums.end());
        
        for(int i=0; i<nums.size()-2; i++){
               int l=i+1, r=nums.size()-1;
                while(l<r){
                    if(nums[i]+nums[l]+nums[r] == 0){
                        if(s.find({nums[i], nums[l], nums[r]}) == s.end()){
                        v.push_back({nums[i], nums[l], nums[r]});
                        s.insert({nums[i], nums[l], nums[r]});
                        }
                        l++;
                    }
                    else if(nums[i]+nums[l]< -nums[r])
                        l++;
                    else
                        r--;
                }
            }
        return v;
    }
};

File: Sequence/q1-q99/q10-q19/q15/q15.py

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        nums.sort()
        solution = []
        
        for i in range(0, len(nums)-2):
            if (not (i > 0 and nums[i] == nums[i-1])):
                left = i + 1
                right = len(nums) -1 
                while left < right:
                    if (nums[left] + nums[right] == -nums[i]):
                        if(nums[left] == nums[right]):
                            solution.append([nums[left], nums[right], nums[i]])
                            break
                        else:
                            solution.append([nums[left], nums[right], nums[i]])
                            while left < right -1 and nums[left] == nums[left+1]:
                                left += 1
                            while left < right -1 and nums[right] == nums[right-1]:
                                right -= 1
                            left += 1
                            right -= 1
                    elif nums[left] + nums[right] > -nums[i]:
                        right -= 1
                    else:
                        left += 1
                
        return solution

